### PR TITLE
Classic EVD naive implementation

### PR DESCRIPTION
**Pull request by RajaSimha**
Implemenetation of the Classic EVD algorithm corresponding to Jacobi rotations in this PR in reference to Algorithm 8.4.2 of the book http://web.mit.edu/ehliu/Public/sclark/Golub%20G.H.,%20Van%20Loan%20C.F.-%20Matrix%20Computations.pdf

Tests are not added yet to the current set of programs.

### CODE REVIEW

File: src/evd.cpp [hunk 1]
```diff
@@ -1,21 +1,85 @@
...
+	}
+	
+	// Store the generated eigen values in the vector
+	for(int i = 0; i < m; i++) {
+		E[i] = A[i*m + i];
+	}
+
+}
 
-void sym_schur2(const double* const X, int n, int p, int q, double* c, double* s) {
```
Comment (ktrianta) [inline src/evd.cpp]: Why is this code deleted? Even if it is not useful, I think it would be better to delete it in a separate commit. Maybe it should be moved to utils?

File: src/evd.cpp [hunk 2]
```diff
@@ -1,21 +1,85 @@
...
-            t = -1.0/(-tau + sqrt(1 + tau*tau));
-        }
-        out_c = 1.0/sqrt(1 + t*t);
-        *c = out_c;
-        *s = t*out_c;
-    } else {
-        *c = 1.0;
-        *s = 0.0;
-    }
+void MatMul(double *P, double *Q, double *R, int n) {
```
Comment (ktrianta) [inline src/evd.cpp]: I think that this code should be in some kind of utils.

File: src/util.cpp [hunk 3]
```diff
@@ -33,6 +33,32 @@ int greater(double x, double y) {
     return x > y;
 }
 
+double sign(double x) {
```
Comment (ktrianta) [inline src/util.cpp]: You can make use of the greater or less function here.

File: src/evd.cpp [hunk 4]
```diff
@@ -1,21 +1,85 @@
-#include <math.h>
-#include "evd.hpp"
+#include<stdio.h>
```
Comment (ktrianta) [inline src/evd.cpp]: Can be removed.

File: src/evd.cpp (hunk 1 above)
Comment (RajaSimha) [reply src/evd.cpp]: The same function has been moved to util.cpp already in the previous set of changes and renamed as `sym_jacobi_coeffs`

File: src/util.cpp (hunk 3 above)
Comment (RajaSimha) [reply src/util.cpp]: Yes, it can be used. I just wanted it to be straightforward for now. Will change it...

File: src/evd.cpp (hunk 2 above)
Comment (RajaSimha) [reply src/evd.cpp]: The same function has been moved into util.cpp in the previous set of changes and renamed as `sym_jacobi_coeffs`

**Comment by eozd** [comment]
I pushed a .clang_format file to the repo previously. We can enforce a coding style by running clang_format program from the root project folder on src files.

File: src/evd.cpp [hunk 5]
```diff
@@ -1,21 +1,84 @@
...
+	for(int i = 0; i < n ; i++) {
+		for(int j = 0; j < n; j++) {
+			for(int k = 0; k < n; k++) {
+				sum += Q[i*n + k] * R[k*n + j];
+			}
+			P[i*n + j] = sum;
+			sum = 0.0;
+		}
+	}
 }
```
Comment (MartinKarp) [inline src/evd.cpp]: To acctualy compute the multiplication is not completely necessary i think, but its fine for the initial implementation.

**Comment by ktrianta** [comment]
Raj could you format the code with clang_format?

File: src/util.cpp (hunk 3 above)
Comment (ktrianta) [reply src/util.cpp]: I was thinking wrong about this.. I think it makes sense to have it as `return (x >= 0.0) ? 1.0 : -1.0`. In case `x == 0.0` the sign should be positive I believe..

File: src/util.hpp [hunk 6]
```diff
@@ -52,6 +52,26 @@ int less(double x, double y);
  */
 int greater(double x, double y);
 
+/**
+ * Return 1 if the number if positive and 0 otherwise.
```
Comment (ktrianta) [inline src/util.hpp]: Is 0 considered positive? Also when the value is negative, is the returned value 0.0 or -1.0?

**Comment by ktrianta** [comment]
> The tests need a revision, they are not correct I think

Could you elaborate on this? It would save us some time if you have already found and can point out the bug.

**Comment by eozd** [comment]
> > The tests need a revision, they are not correct I think
> 
> Could you elaborate on this? It would save us some time if you have already found and can point out the bug.

```evd_test.cpp``` file needs to be added to CMakeLists.txt of the test folder. When I did this, I get compiler errors.

File: test/evd_test.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#include "gtest/gtest.h"
```
Comment (eozd) [inline test/evd_test.cpp]: ```evd_test.cpp``` file needs to be added to tests/CMakeLists.txt together with other test files. Only then it will get compiled.

Comment (MartinKarp) [reply test/evd_test.cpp]: Also the eigenvalues are wrong I think,  I get this in wolfram alpha  
λ_1≈12.7199
λ_2≈5.78305
λ_3≈-5.60024
λ_4≈2.09733

**Comment by MartinKarp** [comment]
> > The tests need a revision, they are not correct I think
> 
> Could you elaborate on this? It would save us some time if you have already found and can point out the bug.

I got these values for the eigenvalues in wolfram alpha. They also correspond better with my implementation.
λ_1≈12.7199
λ_2≈5.78305
λ_3≈-5.60024
λ_4≈2.09733

File: src/evd.cpp (hunk 5 above)
Comment (RajaSimha) [reply src/evd.cpp]: Yes, computing multiplication on the entire matrix is not actually necessary. Since it's a "naive" implementation, I have assumed that there would be no optimization at all at this stage, and the resulting code is just a literal translation of the Rotation Matrix getting multiplied according to the standard algorithm. However, as a first step in the optimizations that we'd perform in the later stages, we can eliminate the Matrix multiplication and replace it with just a vector multiplication (Ref: https://nptel.ac.in/courses/108108079/18) that affects the rows and columns corresponding to the largest off-diagonal element in the update step.

File: src/util.hpp (hunk 6 above)
Comment (RajaSimha) [reply src/util.hpp]: According to what I understand, 0.0 is considered as negative sign for Jacobi rotations.
Ref: http://www2.isec.pt/~armenioc/privado/PapersLivrosSites/Matlab/Cambridge%20University%20Press%20-%20Numerical%20Methods%20in%20Engineering%20with%20MATLAB/Cambridge.University.Press.Numerical.Methods.in.Engineering.with.MATLAB.Aug.2005.eBook-DDU.pdf

Pages : 331 and 332


I'm slightly modifying that function to return an int value instead of double.

File: test/evd_test.cpp (hunk 7 above)
Comment (RajaSimha) [reply test/evd_test.cpp]: Added the entry to the CMakelists.txt. Thanks for pointing it out.

Comment (RajaSimha) [reply test/evd_test.cpp]: Regarding the low precision of the computed values, I think that it's probably due to the slightly different expressions used to compute the rotation matrix, which accumulates and propagates the floating point precision error over the iterations. I'll try to rewrite the expressions with lower number of flops in the subsequent refinements of the program.

File: src/util.hpp (hunk 6 above)
Comment (ktrianta) [reply src/util.hpp]: - Ok! I wil check if also other sources consider sign(x) = -1.0 when x<=0.
- Regarding returning an int, if you are gonna use this 1 or -1 in a multiplication with double, why return an int instead of a double.
- Also the comment should be updated as right now it mentions return values of 1 and 0.

Comment (RajaSimha) [reply src/util.hpp]: Well, most of the implementations of sign() function are designed to return an int, and I followed it assuming that there will be no NaN kind of inputs (Ref: https://stackoverflow.com/a/35491710) and int takes less space and has faster processing speeds than floating points (a kind of optimization if you wanna consider it in that way ;-) ). But it's debatable though.

I've updated the comment too....

Comment (ktrianta) [reply src/util.hpp]: I agree that int takes may be better in some cases but I don't think it is better in our case. Imagine you do `c = sign(a) * b` and `b` and `c` are doubles. The compiler will generate an extra instruction to convert the int returned by `sign(a)` to double.

Comment (RajaSimha) [reply src/util.hpp]: That's true though. We could as well have multiple versions of such utility functions returning different data types, but that becomes too arcane at least for now. Will revert it back to return double.

Comment (eozd) [reply src/util.hpp]: A simple template function should suffice

**Comment by RajaSimha** [comment]
Todo: 
1. Sort the eigenvalues
2. Improve the precision
3. Try to eliminate the "temp" matrix, by using the identity A(i,j) = A(i,j) * [P_t(i) * P(j)]
